main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala

```diff
@@ -0,0 +1,650 @@
+    * For example for the case `Err[Bool]`, we generate:
+    *
+    * public boolean value;
+    * public String tag;
+    *
+    * And for `Error[List[Int32]]` we generate:
+    *
+    * public boolean value;
+    * public String tag;
+    *
+    * @param visitor class visitor
+    * @param name name of the field
+    * @param descriptor descriptor of field
+    */
+  private def compileEnumField(visitor: ClassWriter, name: String, descriptor: String) = {
+    val field = visitor.visitField(ACC_PUBLIC, name, descriptor, null, null)
+    field.visitEnd()
+  }
+
+  /**
+    * Returns descriptor of the field.
+    * if the field is of primitive type it returns it's descriptor, otherwise it returns descriptor of object class.
+    * @param fType type of the `value` field of the class
+    * @param interfaces generated functional interfaces
+    */
+  private def getFieldTypeDescriptor(fType: Option[Type], interfaces: Map[Type, QualName]): String = {
+    fType match {
+      case Some(t) if t.isPrimitive => CodegenHelper.descriptor(t, interfaces)
+      case _ => asm.Type.getDescriptor(Constants.objectClass)
+    }
+  }
+
+  /**
+    * Creates the single argument constructor of the enum case class which is named `qualName`.
+    * The only argument required to instantiate the class is the `value`.
+    * The type of the field of the case is give by `descriptor`
+    * @param visitor class visitor
+    * @param qualName name of the class
+    * @param descriptor descriptor of the `value` field
+    * @param fType type of the `value` field
+    */
+  private def compileEnumConstructor(visitor: ClassWriter, qualName: QualName, descriptor: String, fType: Option[Type]) = {
+    val constructor = visitor.visitMethod(ACC_PUBLIC, "<init>", s"($descriptor)V", null, null)
+    val clazz = Constants.objectClass
+    val ctor = clazz.getConstructor()
+
+    constructor.visitCode()
+    constructor.visitVarInsn(ALOAD, 0)
+
+    // Call the super (java.lang.Object) constructor
+    constructor.visitMethodInsn(INVOKESPECIAL, asm.Type.getInternalName(clazz), "<init>",
+      asm.Type.getConstructorDescriptor(ctor), false)
+
+    // Load instruction for type of `value`
+    val iLoad : Int = fType match {
+      case Some(Type.Var(id, _)) =>  throw InternalCompilerException(s"Non-monomorphed type variable '$id in type '$descriptor'.")
+      case Some(Type.Bool) | Some(Type.Char) | Some(Type.Int8) | Some(Type.Int16) | Some(Type.Int32) => ILOAD
+      case Some(Type.Int64) => LLOAD
+      case Some(Type.Float32) => FLOAD
+      case Some(Type.Float64) => DLOAD
+      case _ => ALOAD
+    }
+
+    // Put the object given to the construct on the `value` field
+    constructor.visitVarInsn(ALOAD, 0)
+    constructor.visitVarInsn(iLoad, 1)
+    constructor.visitFieldInsn(PUTFIELD, decorate(qualName), "value", descriptor)
+
+    // Put the name of the tag on the `tag` field
+    constructor.visitVarInsn(ALOAD, 0)
+    constructor.visitLdcInsn(qualName.ref.last)
```

Comment:
We talked about the problem of relying on `.last`, hopefully something can be done.